src/common/simple-manifest.js

```diff
@@ -0,0 +1,31 @@
+const getSimpleManifest = function (manifest) {
+  // This function generates a simplified manifest with the public and private
+  // variable fields removed. All public and private variables are defined in the
+  // environment field of each container
+
+  const simpleManifest = { ...manifest['manifest'] }
+  delete simpleManifest['vars']
+  const containers = manifest['manifest']['containers']
+  const publicVars = manifest['manifest']['vars']
+  const privateVars = manifest['private'] ? manifest['private']['vars'] : {}
+
+  simpleManifest['containers'] = containers.reduce((agg, container) => {
+    const env = container['environment']
+    const newEnv = {
+      ...env,
+      ...publicVars,
```

Comment:
Rather than expanding the public/private vars into the environment, we should check the values of `env` to see if any vars are used (check if they start with a `$`) and if so, interpolate the correct variable.

Reply:
The idea here is that the interpolation happens while the manifest is being validated, similar to codiusd. It's hard to separate the validation and interpolation steps without duplicating a lot of the checks that are performed on the manifest. This function is only meant to produce a template that is to be filled in with the final env values when validating each container.

Reply:
My latest changes perform the full validation before interpolating the environment. Clears up a lot of these issues.